### PR TITLE
Remove deprecation

### LINKED ISSUE #29: Can't build anymore
running `cargo build --examples` leads to:

```
   Compiling socketcan-alt v0.3.0 (/tmp/tmp.7ZsUJpP56V/socketcan-rs)
error[E0609]: no field `can_dlc` on type `sys::can_frame`
  --> src/frame/data.rs:18:35
   |
18 |             (*inner.as_mut_ptr()).can_dlc = data.len() as _;
   |                                   ^^^^^^^ unknown field
   |
   = note: available fields are: `can_id`, `__bindgen_anon_1`, `__pad`, `__res0`, `len8_dlc`, `data`
help: one of the expressions' fields has a field of the same name
   |
18 |             (*inner.as_mut_ptr()).__bindgen_anon_1.can_dlc = data.len() as _;
   |                                   +++++++++++++++++

error[E0609]: no field `can_dlc` on type `sys::can_frame`
  --> src/frame/data.rs:29:31
   |
29 |         &self.0.data[..self.0.can_dlc as _]
   |                               ^^^^^^^ unknown field
   |
   = note: available fields are: `can_id`, `__bindgen_anon_1`, `__pad`, `__res0`, `len8_dlc`, `data`
help: one of the expressions' fields has a field of the same name
   |
29 |         &self.0.data[..self.0.__bindgen_anon_1.can_dlc as _]
   |                               +++++++++++++++++

error[E0609]: no field `can_dlc` on type `sys::can_frame`
  --> src/frame/remote.rs:18:35
   |
18 |             (*inner.as_mut_ptr()).can_dlc = dlc;
   |                                   ^^^^^^^ unknown field
   |
   = note: available fields are: `can_id`, `__bindgen_anon_1`, `__pad`, `__res0`, `len8_dlc`, `data`
help: one of the expressions' fields has a field of the same name
   |
18 |             (*inner.as_mut_ptr()).__bindgen_anon_1.can_dlc = dlc;
   |                                   +++++++++++++++++

error[E0609]: no field `can_dlc` on type `sys::can_frame`
  --> src/frame/remote.rs:28:16
   |
28 |         self.0.can_dlc
   |                ^^^^^^^ unknown field
   |
   = note: available fields are: `can_id`, `__bindgen_anon_1`, `__pad`, `__res0`, `len8_dlc`, `data`
help: one of the expressions' fields has a field of the same name
   |
28 |         self.0.__bindgen_anon_1.can_dlc
   |                +++++++++++++++++

For more information about this error, try `rustc --explain E0609`.
error: could not compile `socketcan-alt` due to 4 previous errors
```

Seems to be a bindgen issue? :/

### PR DESCRIPTION
**Pull request by hellow554**
according to linux-can/can-utils@c398e56 the `can_dlc` field is deprected now and `len` should be used instead.

Fixes #29

### CODE REVIEW

**Comment by hellow554** [comment]
After merging I would suggest to push a new release

**Comment by hellow554** [comment]
Okay, this is funny and expected.

Should be make a decision in `build.rs` which version of `can.h` is used? Do you have any other ideas? A feature flag maybe?

**Comment by Hakuyume** [comment]
ideas
1. drop old kernel header and update CI to use new kernel header
2. add a feature flag and update CI to use both old/new kernel header
3. do some bindgen hack and update CI to use both old/new kernel header
4. implement automatic switching in build.rs (as @hellow554 mentioned) and update CI to use both old/new kernel header

I think 2 is good and 1 is enough.
My quick survey cannot find good way for 3.
4 seems too complicated.

**Comment by Hakuyume** [comment]
We can do option 1 now. When somebody else want to use old kernel header, I will consider option 2.

**Comment by hellow554** [comment]
I'm probably the one who would like to see option 2, because I'm using it for a custom kernel build with yocto.

If I see that correctly, the changes are happening from version 5.11 on, so anything older than that (e.g. a long term stable kernel 5.10) would be affected by a non-building version.

So, option 2? (:

**Comment by Hakuyume** [comment]
I updated CI in #31. It checks both old/new headers. Can you rebase #30  on it and update the PR to pass the test?

**Comment by Hakuyume** [comment]
For the name of feature flag, how about `can-dlc-unaliased`? If enabled, the code assumes `can_dic` is not a union. (disabled as default)

**Comment by hellow554** [comment]
I think I got it now... let's try it and see.
Maybe we need to adjust the feature flag during building.

Maybe trying to figure it out in the `build.rs` automatically would be the next step? :P

If this is good, I would like to see #31 merged first, then this so I can rebase it again and it will be properly tracked on github.

File: Cargo.toml [hunk 1]
```diff
@@ -13,10 +13,10 @@ exclude = [".github"]
 [package.metadata.docs.rs]
 all-features = true
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
```
Comment (Hakuyume) [inline Cargo.toml:17]: What is your intention?

File: Cargo.toml [hunk 2]
```diff
@@ -13,10 +13,10 @@ exclude = [".github"]
...
 [features]
+default = []
```
Comment (Hakuyume) [inline Cargo.toml]: Do we need this?

File: src/sys.rs [hunk 3]
```diff
@@ -5,3 +5,27 @@
 #![allow(non_snake_case)]
 
 include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+
+pub unsafe fn can_frame_len(frame: *const can_frame) -> u8 {
```
Comment (Hakuyume) [inline src/sys.rs:9]: :+1:

**Comment by Hakuyume** [comment]
> Maybe we need to adjust the feature flag during building.

Yes
I will do that.

File: Cargo.toml (hunk 2 above)
Comment (hellow554) [reply Cargo.toml]: No, but it doesn't hurt? 🤷

File: Cargo.toml (hunk 1 above)
Comment (hellow554) [reply Cargo.toml:17]: Removed it (:

File: src/sys.rs [hunk 4]
```diff
@@ -5,3 +5,27 @@
...
+    #[cfg(feature = "can-dlc-unaliased")]
+    {
+        (*frame).can_dlc.len
```
Comment (Hakuyume) [inline src/sys.rs]:
```suggestion
        (*frame).can_dlc
```

https://github.com/Hakuyume/socketcan-rs/runs/4034470729?check_suite_focus=true#step:6:191

File: src/sys.rs [hunk 5]
```diff
@@ -5,3 +5,27 @@
...
+    #[cfg(not(feature = "can-dlc-unaliased"))]
+    {
+        (*frame).__bindgen_anon_1.len
+    }
+}
+
+pub unsafe fn can_frame_len_mut(frame: *mut can_frame) -> *mut u8 {
+    #[cfg(feature = "can-dlc-unaliased")]
+    {
+        &mut (*frame).can_dlc.len
```
Comment (Hakuyume) [inline src/sys.rs]:
```suggestion
        &mut (*frame).can_dlc
```

https://github.com/Hakuyume/socketcan-rs/runs/4034470729?check_suite_focus=true#step:6:197

**Comment by hellow554** [comment]
Classic copy & paste error... :/

But now we know, that the tests work (:

**Comment by Hakuyume** [comment]
Thank you for your contribution. I merged via #32.

**Pull request merged**